docs/design/enhance-pluggable-policy.md

```diff
@@ -0,0 +1,175 @@
+
+This proposal adds to and extends existing API to so that authorizers may provide the functionality described above.  It does not attempt to describe how the policies themselves can be expressed, that is up the authorization plugins themselves.
+
+
+## Enhancements to existing Authorization interfaces
+The existing Authorization interfaces are described here: [authorization.md](../authorization.md).  A couple additions will allow the development of an Authorizer that matches based on different rules than the existing implementation.
+
+### Request Attributes
+The existing authorizer.Attributes only has 5 attributes (user, groups, isReadOnly, kind, and namespace).  If we add more detailed verbs, content, and resource names, then Authorizer plugins will have the same level of information available to RESTStorage components in order to express more detailed policy.  The replacement excerpt is below.
+
+A request has 7 attributes that can be considered for authorization:
+  - user - the user-string which a user was authenticated as.  This is included in the Context.
+  - groups - the groups to which the user belongs.  This is included in the Context.
+  - verb - string describing the requestion action. Today we have: get, list, watch, create, update, and delete
+  - what kind of object is being accessed 
+    - applies only to the API endpoints, such as 
+        `/api/v1beta1/pods`.  For miscelaneous endpoints, like `/version`, the kind is the empty string.
+  - the namespace of the object being access, or the empty string if the endpoint does not support namespaced objects.  This is included in the Context.
+  - content - the runtime.Object being submitted. This only has content for creates and updates
+  - resourceName - the name of the resource being gotten (get-ed) or deleted.
+
+### Authorizer Interface
+The existing Authorizer interface is very simple, but there isn't a way to provide details about allows, denies, or failures.  The extended detail is useful for UIs that want to describe why certain actions are disallowed.  Not all Authorizers will want to provide that information, but for those that do having that capability is useful.  In addition, adding a `GetAllowedSubjects` method that returns back the users and groups that can perform a particular action makes it possible to answer questions like, "who can see resources in my namespace" (see [ResourceAccessReview](#ResourceAccessReview) further down).
+
+```
+// OLD
+type Authorizer interface {
+  Authorize(a Attributes) error 
+}
+```
+
+```
+// NEW
+type Authorizer interface {
+  // Authorize takes a Context (for namespace, user, and traceability) and Attributes to make a policy determination.
+  // reason is an optional return value that can describe why a policy decision was made.
+  Authorize(ctx api.Context, a Attributes) (allowed bool, reason string, evaluationError error)
```

Comment:
@derekwaynecarr Isn't this addressed in the passage that describes UIs that want to display reason information?